sync3/lists.go

```diff
@@ -138,6 +138,18 @@ func (s *InternalRequestLists) Get(listKey string) *FilteredSortableRooms {
 	return s.lists[listKey]
 }
 
+// SnapshotRoomIDs builds a map from list name to the room IDs in that list's sliding
+// window. The result is safe to modify by the caller.
+func (s *InternalRequestLists) SnapshotRoomIDs() map[string][]string {
+	roomIDsByList := make(map[string][]string, s.Len())
+	for listName, listData := range s.lists {
+		roomsInList := make([]string, len(listData.roomIDs))
```

Comment:
`listData.roomIDs` is every room in the entire filtered list (e.g all DM rooms if `is_dm: true`), NOT the rooms visible in the sliding window. See https://github.com/matrix-org/sliding-sync/blob/c4b7dcb610d112fa7c094a6f322b2cbd4a3b57e4/sync3/sort.go#L49

You need to pass through the `muxedReq` ranges for each list to know which indexes to include.